### PR TITLE
cmd/tailscale/web: restrict web access to synology admins.

### PR DESCRIPTION
**Pull request by maisem**
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

**Review by crawshaw — APPROVED** [review]

File: ipn/ipnserver/server.go [hunk 1]
```diff
@@ -320,6 +321,12 @@ func isReadonlyConn(ci connIdentity, operatorUID string, logf logger.Logf) bool
 		// has a different last-user-wins auth model.
 		return false
 	}
+	// TODO(maisem): what should we do here? This is the easiest option to allow
```
Comment (bradfitz) [inline ipn/ipnserver/server.go:324]: tailscaled should authenticate that the connection came from the web/CGI script, and the web/CGI shoudl authenticate the client HTTP request came came from a Synology admin using Synology's cookie auth stuff.

Comment (maisem) [reply ipn/ipnserver/server.go:324]: Right now we don't restrict access to Synology admins. I know some Synology apps do that, but I have no idea how to configure that. Will dig into this.

Comment (maisem) [reply ipn/ipnserver/server.go:324]: Minor digging suggests that we can actually not restrict web access to only admins https://old.reddit.com/r/synology/comments/8bvjkz/application_permissions/dxadkvv/

Comment (crawshaw) [reply ipn/ipnserver/server.go:324]: Can we not block getting the next version of the synology package delivered to synology for review?

Comment (maisem) [reply ipn/ipnserver/server.go]: @bradfitz thanks for the suggestion to use `/etc/group`, I think this works.

File: cmd/tailscale/cli/web.go [hunk 2]
```diff
@@ -82,17 +84,57 @@ func runWeb(ctx context.Context, args []string) error {
...
-	if distro.Get() == distro.Synology {
-		cmd := exec.Command("/usr/syno/synoman/webman/modules/authenticate.cgi")
-		out, err := cmd.CombinedOutput()
-		if err != nil {
-			return "", fmt.Errorf("auth: %v: %s", err, out)
+// authorize checks whether the provided user has access to the web UI.
+// On Synology, it does this verification by consulting the membership of the
+// "administrators" group.
+func authorize(name string) error {
+	if distro.Get() != distro.Synology {
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:91]: QNAP is coming soon, so maybe:

```go
if distro.Get() == disto.Synology {
     return authorizeSynology(name)
}
return nil
```

File: cmd/tailscale/cli/web.go [hunk 3]
```diff
@@ -82,17 +84,57 @@ func runWeb(ctx context.Context, args []string) error {
...
+	for s.Scan() {
+		l := s.Text()
+		if !strings.HasPrefix(l, "administrators:") {
+			continue
 		}
-		return string(out), nil
+		agLine = l
+		break
+	}
+	if agLine == "" {
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:109]: if err := bs.Err(); err != nil {
      ...

File: cmd/tailscale/cli/web.go [hunk 4]
```diff
@@ -82,17 +84,57 @@ func runWeb(ctx context.Context, args []string) error {
...
+	f, err := os.Open("/etc/group")
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	s := bufio.NewScanner(f)
+	var agLine string
+	for s.Scan() {
+		l := s.Text()
+		if !strings.HasPrefix(l, "administrators:") {
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:103]: This way doesn't allocate, FWIW:

```
import "go4.org/mem"

if !mem.HasPrefix(mem.B(s.Bytes()), mem.S("administrators:") {
   continue
 }
agLine = s.Text()
```

**Review by bradfitz — APPROVED** [review]

**Pull request merged**